README.md
# {splitTools} <a href='https://github.com/mayer79/splitTools'><img src='man/figures/logo.png' align="right" height="139" /></a>

<!-- badges: start -->

[![CRAN status](http://www.r-pkg.org/badges/version/splitTools)](https://cran.r-project.org/package=splitTools)
[![R-CMD-check](https://github.com/mayer79/splitTools/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/mayer79/splitTools/actions)
[![Codecov test coverage](https://codecov.io/gh/mayer79/splitTools/branch/main/graph/badge.svg)](https://app.codecov.io/gh/mayer79/splitTools?branch=main)

[![](https://cranlogs.r-pkg.org/badges/splitTools)](https://cran.r-project.org/package=splitTools) 
[![](https://cranlogs.r-pkg.org/badges/grand-total/splitTools?color=orange)](https://cran.r-project.org/package=splitTools)

<!-- badges: end -->

## Overview

{splitTools} is a toolkit for fast data splitting. It does not have any dependencies. 

Its two main functions `partition()` and `create_folds()` support

- data partitioning (e.g. into training, validation and test),
- creating (in- or out-of-sample) folds for cross-validation (CV),
- creating *repeated* folds for CV,
- stratified splitting, 
- grouped splitting as well as
- blocked splitting (if the sequential order of the data should be retained).

The function `create_timefolds()` does time-series splitting where the out-of-sample data follows the (extending or moving) in-sample data.

The result of `create_folds()` can be directly passed to the `folds` argument in CV functions of XGBoost or LightGBM. Since these functions expect *out-of-sample* indices, set the option `invert = TRUE`.

## Installation

```r
# From CRAN
install.packages("splitTools")

# Development version
devtools::install_github("mayer79/splitTools")
```

## Usage

``` r
library(splitTools)

p <- c(train = 0.5, valid = 0.25, test = 0.25)

# Train/valid/test indices for iris data stratified by Species
str(inds <- partition(iris$Species, p, seed = 1))

# List of 3
#  $ train: int [1:73] 1 3 5 7 8 10 12 13 14 15 ...
#  $ valid: int [1:38] 4 9 19 21 27 28 29 30 32 35 ...
#  $ test : int [1:39] 2 6 11 16 18 22 26 37 38 40 ...

# Same, but different output interface
head(inds <- partition(iris$Species, p, split_into_list = FALSE, seed = 1))

# [1] train test  train valid train test 
# Levels: train valid test

# In-sample indices for 5-fold CV (stratified by Species)
str(inds <- create_folds(iris$Species, k = 5, seed = 1))

# List of 5
#  $ Fold1: int [1:120] 2 4 5 6 7 8 9 10 11 15 ...
#  $ Fold2: int [1:120] 1 2 3 4 5 6 9 10 11 12 ...
#  $ Fold3: int [1:120] 1 2 3 4 6 7 8 9 11 12 ...
#  $ Fold4: int [1:120] 1 3 5 6 7 8 10 11 12 13 ...
#  $ Fold5: int [1:120] 1 2 3 4 5 7 8 9 10 12 ...

# In-sample indices for 3 times repeated 5-fold CV (stratified by Species)
str(inds <- create_folds(iris$Species, k = 5, m_rep = 3, seed = 1))

# List of 15
#  $ Fold1.Rep1: int [1:120] 2 4 5 6 7 8 9 10 11 15 ...
#  $ Fold2.Rep1: int [1:120] 1 2 3 4 5 6 9 10 11 12 ...
#  $ Fold3.Rep1: int [1:120] 1 2 3 4 6 7 8 9 11 12 ...
#  $ Fold4.Rep1: int [1:120] 1 3 5 6 7 8 10 11 12 13 ...
#  $ Fold5.Rep1: int [1:120] 1 2 3 4 5 7 8 9 10 12 ...
#  $ Fold1.Rep2: int [1:120] 1 2 3 4 5 6 8 9 11 12 ...
#  $ Fold2.Rep2: int [1:120] 1 3 6 7 8 9 10 12 13 14 ...
# [...]

# Indices for time-series splitting
str(inds <- create_timefolds(1:100, k = 5))

# List of 5
# $ Fold1:List of 2
#  ..$ insample : int [1:17] 1 2 3 4 5 6 7 8 9 10 ...
#  ..$ outsample: int [1:17] 18 19 20 21 22 23 24 25 26 27 ...
# $ Fold2:List of 2
#  ..$ insample : int [1:34] 1 2 3 4 5 6 7 8 9 10 ...
#  ..$ outsample: int [1:17] 35 36 37 38 39 40 41 42 43 44 ...
# $ Fold3:List of 2
# [...]
```

For more details, check out the vignette.


NEWS.md
# splitTools 1.0.1

- Documentation update.
- Remove dependency {ranger}.

# splitTools 1.0.0

- Documentation update
- Push the package to mature version 1.0.0.
- Replacing `importFrom` by `::`.

# splitTools 0.3.3

## New Functionality

- Sometimes, one need to stratify on multiple columns. The new function `multi_strata()` provides a vector of stratification groups based on a data frame that can be then passed to `partition()` or `create_folds()`. Each stratification group will contain "similar" data rows, where similarity is either based on a kmeans cluster analysis or forming all combinations of binned columns. Thanks to [kapsner](https://github.com/kapsner) for the idea and the help with the implementation.

## Maintenance

- Set up github actions, thanks to L. Kapsner.

# splitTools 0.3.2

This is documentation and maintenance update only with the following changes:

- Updated documentation to clarify that `create_folds()` creates in-sample indices by default. If out-of-sample indices are to be generated, set `invert = TRUE`.
- Got rid of a CRAN check notes about LazyData.
- Changed to better way of updating/generating the package.

# splitTools 0.3.1

## New Functionality

- `create_folds` and `partition` have received a `shuffle` option to shuffle rows within folds/partitions. The default is FALSE.

# splitTools 0.3.0

## Breaking change for tiny data sets

- `create_folds` and `partition` cannot return empty folds/partitions anymore. This impacts only extremely small data sets.

## Other

- Unit tests have been added.

# splitTools 0.2.1

## New Functionality

- `create_timefolds` now allows also moving windows training data, not just extending windows data.

## Other

Reduced minimally required R version from 3.5 to 3.1.

# splitTools 0.2.0

## New Functionality

- Added `type = "blocked"` to `create_folds` and `partition` to allow for blocked splitting.

- Added function `create_timefolds` for cross-validation of time series data.

## Other

Added vignette

# splitTools 0.1.0

This is the initial CRAN release.


tests/testthat.R
library(testthat)
library(splitTools)

test_check("splitTools")


tests/testthat/test-partition.R
test_that("number of partitions is correct", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- partition(y, p = c(0.6, 0.2, 0.2))
  expect_equal(length(out), 3L)
})

test_that("no observation is lost", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- partition(y, p = c(0.6, 0.2, 0.2))
  expect_equal(sort(unique(unlist(out))), seq_along(y))
})

test_that("no observation is doubled", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- partition(y, p = c(0.6, 0.2, 0.2))
  expect_equal(length(unlist(out)), length(y))
})

test_that("partitions are exact for integer n*p and basic splitting", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.6, 0.2, 0.2)
  out <- partition(y, p, type = "basic")
  expect_equal(vapply(out, length, FUN.VALUE = 1, USE.NAMES = FALSE), length(y) * p)
})

test_that("stratified splitting is balanced for integer n*p within stratum", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.6, 0.2, 0.2)
  out <- partition(y, p, type = "stratified")
  expect_equal(length(unique(table(y[out[[1]]]))), 1)
})

test_that("argument 'split_into_list' works", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.6, 0.2, 0.2)
  expect_true(is.list(partition(y, p, split_into_list = TRUE)))
  expect_false(is.list(partition(y, p, split_into_list = FALSE)))
})

test_that("argument 'seed' works", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.6, 0.2, 0.2)
  expect_equal(partition(y, p, seed = 1), partition(y, p, seed = 1))
})

test_that("argument 'shuffle' works for stratified splitting", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.6, 0.2, 0.2)
  out1 <- partition(y, p, seed = 1)
  out2 <- partition(y, p, seed = 1, shuffle = TRUE)
  expect_false(all(out1[[1]] == out2[[1]]))
  expect_true(all(out1[[1]] == sort(out2[[1]])))
})

test_that("argument 'shuffle' works for grouped splitting", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.6, 0.2, 0.2)
  out1 <- partition(y, p, seed = 1, type = "grouped")
  out2 <- partition(y, p, seed = 1, shuffle = TRUE, type = "grouped")
  expect_false(all(out1[[1]] == out2[[1]]))
  expect_true(all(out1[[1]] == sort(out2[[1]])))
})

test_that("argument 'shuffle' raises message with split_into_list = FALSE", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.6, 0.2, 0.2)
  expect_message( partition(y, p,  split_into_list = FALSE, shuffle = TRUE))
})

test_that("argument 'use_names' works", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.6, 0.2, 0.2)
  expect_equal(names(partition(y, p, use_names = TRUE)),
               names(partition(y, p, use_names = FALSE)))
  names(p) <- c("a", "b", "c")
  expect_named(partition(y, p, use_names = TRUE), c("a", "b", "c"))
  expect_named(partition(y, p, use_names = FALSE), c("1", "2", "3"))
})

test_that("blocked splitting works", {
  y <- rep(LETTERS[1:10], each = 10)
  p <- c(0.5, 0.5)
  out <- partition(y, p, type = "blocked")
  expect_equal(out[[1]], 1:50)
  expect_equal(out[[2]], 51:100)
})

test_that("grouped splitting works", {
  y <- rep(LETTERS[1:2], each = 50)
  p <- c(0.5, 0.5)
  out <- partition(y, p, type = "grouped")
  expect_true(all(out[[1]] == 1:50) || all(out[[2]] == 1:50))
})

test_that("basic splitting works with seed", {
  y <- rep(LETTERS[1:10], each = 5)
  p <- c(0.5, 0.5)
  out <- partition(y, p, type = "basic", seed = 1)
  expect_equal(out[[1]][1:7], c(1, 3, 5, 7, 8, 10, 12))
})

test_that("stratified splitting works with seed", {
  y <- rep(LETTERS[1:10], each = 5)
  p <- c(0.5, 0.5)
  out <- partition(y, p, type = "stratified", seed = 1)
  expect_equal(out[[1]][1:7], c(1, 3, 5, 7, 8, 10, 11))
})

test_that("grouped splitting works with seed", {
  y <- rep(LETTERS[1:10], each = 5)
  p <- c(0.5, 0.5)
  out <- partition(y, p, type = "grouped", seed = 1)
  expect_equal(out[[1]][1:10], c(6:10, 16:20))
})

test_that("character and factor input give same result", {
  y <- rep(LETTERS[1:10], each = 5)
  p <- c(0.5, 0.5)
  out1 <- partition(y, p, type = "grouped", seed = 1)
  out2 <- partition(factor(y), p, type = "grouped", seed = 1)
  expect_equal(out1, out2)
})

test_that("stratified splitting on continous data works", {
  y <- seq(0, 1, by = 0.01)
  p <- c(0.5, 0.5)
  out <- partition(y, p, n_bins = 2, seed = 1)
  expect_lt(abs(diff(sapply(out, mean))), 1)
})

test_that("stratified splitting on continous data works with missing values", {
  y <- c(1:99, NA)
  p <- c(0.5, 0.5)
  out <- partition(y, p, n_bins = 2, seed = 1)
  expect_equal(length(unlist(out)), length(y))
})

test_that("stratified splitting on continous data reacts on n_bins", {
  y <- rep(1:10, each = 10)
  p <- c(0.5, 0.5)
  out1 <- partition(y, p, n_bins = 2, seed = 1)
  out2 <- partition(y, p, n_bins = 11, seed = 1)
  expect_equal(isTRUE(all.equal(out1, out2)), FALSE)
})

test_that("tiny data sets are providing non-empty partitions for stratified sampling", {
  y <- c("A", "A", "B")
  expect_message(partition(y, p = c(0.9, 0.1), seed = 1))
  expect_message(partition(y, p = c(0.9, 0.1, 0.2), seed = 10))
  suppressMessages(expect_message(partition(1:2, p = c(0.3, 0.3, 0.4), seed = 4)))
})

test_that("tiny data sets are providing non-empty partitions for basic sampling", {
  y <- c("A", "A", "B")
  expect_message(partition(y, p = c(0.9, 0.1), seed = 2, type = "basic"))
  expect_message(partition(y, p = c(0.9, 0.1, 0.2), seed = 10, type = "basic"))
})

test_that("tiny data sets are providing non-empty partitions for grouped sampling", {
  y <- c("A", "A", "B")
  expect_message(partition(y, p = c(0.9, 0.1), seed = 1, type = "grouped"))
})





tests/testthat/test-create_folds.R
test_that("number of folds is correct", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- create_folds(y, k = 3)
  expect_equal(length(out), 3L)
})

test_that("no observation is lost", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- create_folds(y, k = 4)
  expect_equal(sort(unique(unlist(out))), seq_along(y))
})

test_that("no observation is doubled", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- create_folds(y, k = 4)
  expect_equal(length(unlist(out)), length(y) * 3)
})

test_that("folds have same size for integer n/k and basic splitting", {
  y <- rep(LETTERS[1:10], each = 10)
  k <- 4
  out <- create_folds(y, k = k, type = "basic")
  expect_equal(vapply(out, length, FUN.VALUE = 1, USE.NAMES = FALSE),
               rep(length(y) - length(y) / k, k))
})

test_that("stratified splitting is balanced for integer n/k within stratum", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- create_folds(y, k = 2, type = "stratified")
  expect_equal(length(unique(table(y[out[[1]]]))), 1)
})

test_that("argument 'invert' works", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- create_folds(y, k = 4, invert = TRUE)
  expect_equal(length(unlist(out)), length(y))
})

test_that("argument 'seed' works", {
  y <- rep(LETTERS[1:10], each = 10)
  expect_equal(create_folds(y, k = 3, seed = 1),
               create_folds(y, k = 3, seed = 1))
})

test_that("argument 'use_names' works", {
  y <- rep(LETTERS[1:10], each = 10)
  expect_named(create_folds(y, k = 3, use_names = TRUE), c("Fold1", "Fold2", "Fold3"))
  expect_named(create_folds(y, k = 3, use_names = FALSE), NULL)
})

test_that("argument 'shuffle' works for stratified splitting", {
  y <- rep(LETTERS[1:10], each = 10)
  out1 <- create_folds(y, k = 3, seed = 1)
  out2 <- create_folds(y, k = 3, seed = 1, shuffle = TRUE)
  expect_false(all(out1[[1]] == out2[[1]]))
  expect_true(all(out1[[1]] == sort(out2[[1]])))
})

test_that("argument 'shuffle' works for grouped splitting", {
  y <- rep(LETTERS[1:10], each = 10)
  out1 <- create_folds(y, k = 3, seed = 1, type = "grouped")
  out2 <- create_folds(y, k = 3, seed = 1, shuffle = TRUE, type = "grouped")
  expect_false(all(out1[[1]] == out2[[1]]))
  expect_true(all(out1[[1]] == sort(out2[[1]])))
})

test_that("argument 'm_rep' works", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- create_folds(y, k = 5, m_rep = 2)
  expect_equal(length(out), 2 * 5)
})

test_that("blocked splitting works", {
  y <- rep(LETTERS[1:10], each = 10)
  out <- create_folds(y, k = 4, type = "blocked", invert = TRUE)
  expect_equal(out[[1]], 1:25)
})

test_that("grouped splitting works", {
  y <- rep(LETTERS[1:2], each = 50)
  out <- create_folds(y, k = 2, type = "grouped", invert = TRUE)
  expect_true(all(out[[1]] == 1:50) || all(out[[2]] == 1:50))
})

test_that("basic splitting works with seed", {
  y <- rep(LETTERS[1:10], each = 5)
  out <- create_folds(y, k = 3, type = "basic", seed = 1)
  expect_equal(out[[1]][1:7], c(2, 4, 5, 6, 8, 9, 10))
})

test_that("stratified splitting works with seed", {
  y <- rep(LETTERS[1:10], each = 5)
  out <- create_folds(y, k = 3, type = "stratified", seed = 1, invert = TRUE)
  expect_equal(out[[1]][1:7], c(1, 5, 8, 10, 12, 15, 17))
})

test_that("grouped splitting works with seed", {
  y <- rep(LETTERS[1:10], each = 5)
  out <- create_folds(y, k = 3, type = "grouped", seed = 1)
  expect_equal(out[[1]][1:10], c(1:5, 11:15))
})

test_that("character and factor input give same result", {
  y <- rep(LETTERS[1:10], each = 5)
  k <- 3
  out1 <- create_folds(y, k = k, type = "grouped", seed = 1)
  out2 <- create_folds(factor(y), k = k, type = "grouped", seed = 1)
  expect_equal(out1, out2)
})

test_that("stratified splitting on continous data works with missing values", {
  y <- c(1:99, NA)
  out <- create_folds(y, k = 2, seed = 1)
  expect_equal(length(unlist(out)), length(y))
})

test_that("stratified splitting on continous data reacts on n_bins", {
  y <- rep(1:10, each = 10)
  out1 <- create_folds(y, k = 2, n_bins = 2, seed = 1)
  out2 <- create_folds(y, k = 2, n_bins = 50, seed = 1)
  expect_equal(isTRUE(all.equal(out1, out2)), FALSE)
  expect_equal(diff(vapply(out2, function(z) mean(y[z]),
                           FUN.VALUE = numeric(1), USE.NAMES = FALSE)), 0)
})

test_that("tiny data sets are providing non-empty folds for stratified sampling", {
  y <- c("A", "A", "B")
  expect_message(create_folds(y, k = 3, seed = 4))
})

test_that("tiny data sets are providing non-empty folds for basic sampling", {
  y <- c("A", "A", "B", "B")
  expect_message(create_folds(y, k = 3, seed = 4, type = "basic"))
})



R/create_folds.R
#' Create Folds
#'
#' This function provides a list of row indices used for k-fold cross-validation
#' (basic, stratified, grouped, or blocked). Repeated fold creation is supported as well.
#' By default, in-sample indices are returned.
#'
#' By default, the function uses stratified splitting. This will balance the folds
#' regarding the distribution of the input vector `y`.
#' (Numeric input is first binned into `n_bins` quantile groups.)
#' If `type = "grouped"`, groups specified by `y` are kept together
#' when splitting. This is relevant for clustered or panel data.
#' In contrast to basic splitting, `type = "blocked"` does not sample
#' indices at random, but rather keeps them in sequential groups.
#'
#' @inheritParams partition
#' @param k Number of folds.
#' @param m_rep How many times should the data be split into k folds?
#'   Default is 1, i.e., no repetitions.
#' @param use_names Should folds be named? Default is `TRUE`.
#' @param invert Set to `TRUE` in order to receive out-of-sample indices.
#'   Default is `FALSE`, i.e., in-sample indices are returned.
#' @param shuffle Should row indices be randomly shuffled within folds?
#'   Default is `FALSE`.
#' @returns If `invert = FALSE` (the default), a list with in-sample row indices.
#'   If `invert = TRUE`, a list with out-of-sample indices.
#' @export
#' @examples
#' y <- rep(c(letters[1:4]), each = 5)
#' create_folds(y)
#' create_folds(y, k = 2)
#' create_folds(y, k = 2, m_rep = 2)
#' create_folds(y, k = 3, type = "blocked")
#' @seealso [partition()], [create_timefolds()]
create_folds <- function(y, k = 5L,
                         type = c("stratified", "basic", "grouped", "blocked"),
                         n_bins = 10L, m_rep = 1L, use_names = TRUE,
                         invert = FALSE, shuffle = FALSE, seed = NULL) {
  # Input checks
  type <- match.arg(type)
  stopifnot(
    is.atomic(y),
    length(y) >= 2L,
    k >= 2L,
    m_rep >= 1L
  )
  if (type == "blocked" && m_rep > 1L) {
    message("Repeated blocked cross-validation will return the same indices for each repetition. Setting m_rep = 1...")
    m_rep <- 1L
  }

  # Initializations
  if (!is.null(seed)) {
    set.seed(seed)
  }
  p <- rep(1 / k, times = k)
  if (use_names && m_rep > 1L) {
    sfx <- .names("Rep", seq_len(m_rep))
  }
  f <- function(i = 1L) {
    res <- partition(
      y = y,
      p = p,
      type = type,
      n_bins = n_bins,
      split_into_list = TRUE,
      use_names = FALSE,
      shuffle = shuffle
    )
    if (!invert) {
      if (shuffle) {
        res <- lapply(seq_along(res), function(i) unlist(res[-i], use.names = FALSE))
      } else {
        res <- lapply(res, function(z) seq_along(y)[-z])
      }
    }
    if (use_names) {
      names(res) <- .names("Fold", seq_along(res))
      if (m_rep > 1L) {
        names(res) <- paste(names(res), sfx[i], sep = ".")
      }
    } else {
      res <- unname(res)
    }
    res
  }

  # Call partition once or multiple times
  if (m_rep == 1) f() else unlist(lapply(seq_len(m_rep), f), recursive = FALSE)
}

# Little helper(s)
.names <- function(prefix, suffix) {
  paste0(prefix, gsub(" ", "0", format(suffix)))
}



R/multi_strata.R
#' Create Strata from Multiple Features
#'
#' @description
#' Creates a stratification vector based on multiple columns of
#' a `data.frame` that can then be passed to the splitting functions.
#'
#' Currently, the function offers two strategies to create the strata:
#' - "kmeans": k-means cluster analysis on scaled input. (Ordered factors are integer
#'   encoded first, unordered factors and character columns are one-hot-encoded.)
#' - "interaction": All combinations (after binning numeric columns into
#'   approximately `k` bins).
#'
#' @param df A `data.frame` used to form the stratification vector.
#' @param strategy A string (either "kmeans" or "interaction") to compute the strata,
#'   see description.
#' @param k An integer. For `strategy = "kmeans"`, it is the desired number of strata,
#'   while for `strategy = "interaction"`, it is the approximate number of bins per
#'   numeric feature before forming all combinations.
#' @returns Factor with strata as levels.
#' @export
#' @examples
#' y_multi <- data.frame(
#'   A = rep(c(letters[1:4]), each = 20),
#'   B = factor(sample(c(0, 1), 80, replace = TRUE)),
#'   c = rnorm(80)
#' )
#' y <- multi_strata(y_multi, k = 3)
#' folds <- create_folds(y, k = 5)
#' @seealso [partition()], [create_folds()]
multi_strata <- function(df, strategy = c("kmeans", "interaction"), k = 3L) {
  strategy <- match.arg(strategy)
  stopifnot(is.data.frame(df), k >= 2L, k <= nrow(df))
  FUN <- switch(strategy, "kmeans" = .kmeans, "interaction" = .interaction)
  FUN(.good_cols(df), k = k)
}

#=======================
# HELPER FUNCTIONS
#=======================

# Strategy: kmeans
.kmeans <- function(df, k) {
  # Treat ordered as numeric
  v <- colnames(df)[vapply(df, is.ordered, FUN.VALUE = logical(1L))]
  if (length(v) >= 1L) {
    df[v] <- lapply(df[v], as.integer)
  }

  # Now the real work
  df <- scale(stats::model.matrix(~ . + 0, data = df))
  factor(stats::kmeans(df, centers = k)$cluster)
}

# Strategy: interactions across all reasonable columns
.interaction <- function(df, k) {
  v <- colnames(df)[vapply(df, is.numeric, FUN.VALUE = logical(1L))]
  if (length(v) >= 1L) {
    df[v] <- lapply(df[v], .bin_pretty, n_bins = k)
  }
  interaction(df, drop = TRUE, sep = ":")
}

# Select reasonable columns
.good_cols <- function(x) {
  ok <- vapply(
    x,
    function(v)
      is.factor(v) || is.character(v) || is.numeric(v) || is.logical(v),
    FUN.VALUE = logical(1L)
  )
  v <- colnames(x)[ok]
  if (length(v) == 0L) {
    stop("No numeric, factor, character or logical columns in data")
  }
  x[v]
}

# Cuts x into quantile groups (like .bin, but with nice labels)
.bin_pretty <- function(x, n_bins) {
  # +1 required as e.g. cutting a vector with breaks of length 5 results in
  # 4 groups and users define with 'num_cat' the number of categories
  # for numeric variables
  probs <- seq(0, 1, len = n_bins + 1L)
  breaks <- unique(stats::quantile(x, probs = probs, names = FALSE))
  cut(x, breaks, include.lowest = TRUE)
}


R/create_timefolds.R
#' Creates Folds for Time Series Data
#'
#' This function provides a list with in- and out-of-sample indices per fold used for
#' time series k-fold cross-validation, see Details.
#'
#' The data is first partitioned into \eqn{k+1} sequential blocks \eqn{B_1} to
#' \eqn{B_{k+1}}. Each fold consists of two index vectors: one with in-sample row numbers,
#' the other with out-of-sample row numbers. The first fold uses \eqn{B_1} as in-sample
#' and \eqn{B_2} as out-of-sample data. The second one uses either \eqn{B_2}
#' (if `type = "moving"`) or \eqn{\{B_1, B_2\}} (if `type = "extending"`)
#' as in-sample, and \eqn{B_3} as out-of-sample data etc. Finally, the kth fold uses
#' \eqn{\{B_1, ..., B_k\}} ("extending") or \eqn{B_k} ("moving") as in-sample data,
#' and \eqn{B_{k+1}} as out-of-sample data. This makes sure that out-of-sample data
#' always follows in-sample data.
#'
#' @param y Any vector of the same length as the data intended to split.
#' @param k Number of folds.
#' @param use_names Should folds be named? Default is `TRUE`.
#' @param type Should in-sample data be "extending" over the folds (default)
#'   or consist of one single fold ("moving")?
#' @returns
#'   A nested list with in-sample and out-of-sample indices per fold.
#' @export
#' @examples
#' y <- runif(100)
#' create_timefolds(y)
#' create_timefolds(y, use_names = FALSE)
#' create_timefolds(y, use_names = FALSE, type = "moving")
#' @seealso [partition()], [create_folds()]
create_timefolds <- function(y, k = 5L, use_names = TRUE,
                             type = c("extending", "moving")) {
  # Input checks
  type <- match.arg(type)
  stopifnot(is.atomic(y), length(y) >= 2L, k >= 2L)

  # Blocked partition of k + 1 blocks
  res <- partition(
    y = y,
    p = rep(1 / (k + 1L), times = k + 1L),
    type = "blocked",
    use_names = FALSE
  )

  # Create k in-/outsamples for given type
  if (type == "extending") {
    res <- lapply(res[-1L], .make_in_outsample_e)
  } else if (type == "moving") {
    res <- lapply(seq_len(k), .make_in_outsample_m, res)
  }

  # Organize output
  if (use_names) {
    # .names defined in "create_folds.R"
    names(res) <- .names("Fold", seq_along(res))
  } else {
    res <- unname(res)
  }
  res
}

# Helper functions
.make_in_outsample_e <- function(z) {
  list(insample = seq_len(z[1L] - 1L), outsample = z)
}

.make_in_outsample_m <- function(i, res) {
  stats::setNames(res[c(i, i + 1L)], c("insample", "outsample"))
}



R/partition.R
#' Split Data into Partitions
#'
#' This function provides row indices for data splitting, e.g., to split data
#' into training, validation, and test. Different types of split strategies are
#' supported, see Details.
#' The partition indices are either returned as list with one element per partition
#' (the default) or as vector of partition IDs.
#'
#' By default, the function uses stratified splitting. This will balance the partitions
#' as good as possible regarding the distribution of the input vector `y`.
#' (Numeric input is first binned into `n_bins` quantile groups.)
#' If `type = "grouped"`, groups specified by `y` are kept together when
#' splitting. This is relevant for clustered or panel data.
#' In contrast to basic splitting, `type = "blocked"` does not sample indices
#' at random, but rather keeps them in groups: e.g., the first 80% of observations form
#' a training set and the remaining 20% are used for testing.
#'
#' @param y Either the variable used for "stratification" or "grouped" splits.
#'   For other types of splits, any vector of the same length as the data
#'   intended to split.
#' @param p A vector with split probabilities per partition, e.g.,
#'   `c(train = 0.7, valid = 0.3)`. Names are passed to the output.
#' @param type Split type. One of "stratified" (default), "basic", "grouped", "blocked".
#' @param n_bins Approximate numbers of bins for numeric `y`
#'   (only for `type = "stratified"`).
#' @param split_into_list Should the resulting partition vector be split into a list?
#'   Default is `TRUE`.
#' @param use_names Should names of `p` be used as partition names?
#'   Default is `TRUE`.
#' @param shuffle Should row indices be randomly shuffled within partition?
#'   Default is `FALSE`. Shuffling is only possible when `split_into_list = TRUE`.
#' @param seed Integer random seed.
#' @returns
#'   A list with row indices per partition (if `split_into_list = TRUE`)
#'   or a vector of partition IDs.
#' @export
#' @examples
#' y <- rep(c(letters[1:4]), each = 5)
#' partition(y, p = c(0.7, 0.3), seed = 1)
#' partition(y, p = c(0.7, 0.3), split_into_list = FALSE, seed = 1)
#' p <- c(train = 0.8, valid = 0.1, test = 0.1)
#' partition(y, p, seed = 1)
#' partition(y, p, split_into_list = FALSE, seed = 1)
#' partition(y, p, split_into_list = FALSE, use_names = FALSE, seed = 1)
#' partition(y, p = c(0.7, 0.3), type = "grouped")
#' partition(y, p = c(0.7, 0.3), type = "blocked")
#' @seealso [create_folds()]
partition <- function(y, p,
                      type = c("stratified", "basic", "grouped", "blocked"),
                      n_bins = 10L, split_into_list = TRUE, use_names = TRUE,
                      shuffle = FALSE, seed = NULL) {
  # Input checks
  type <- match.arg(type)
  stopifnot(
    length(p) >= 1L,
    p > 0,
    is.atomic(y),
    (n <- length(y)) >= 2L
  )

  # Initializations
  p <- p / sum(p)
  if (!is.null(seed)) {
    set.seed(seed)
  }

  # Calculation of partition ids
  if (type == "basic") {
    out <- .smp_fun(n, p)
    out <- .fill_empty_partitions(out, p = p)
  } else if (type == "blocked") {
    out <- rep.int(seq_along(p), times = ceiling(p * n))[seq_len(n)]
  } else if (type == "stratified") {
    if (is.numeric(y) && length(unique(y)) > n_bins) {
      y <- .bin(y, n_bins)
    }
    if (anyNA(y)) {
      y <- factor(y, exclude = NULL)
    }
    out <- stats::ave(integer(n), y, FUN = function(z) .smp_fun(length(z), p))
    out <- .fill_empty_partitions(out, p = p)
  } else if (type == "grouped") {
    y_unique <- unique(y)
    m <- length(y_unique)
    stopifnot(length(p) <= m)
    y_folds <- .smp_fun(m, p)
    y_folds <- .fill_empty_partitions(y_folds, p = p)
    out <- y_folds[match(y, y_unique)]
  }

  # Output
  if (use_names && !is.null(names(p))) {
    out <- factor(out, levels = seq_along(p), labels = names(p))
  }
  if (!split_into_list) {
    if (shuffle) {
      message("Shuffling has no effect with split_into_list = TRUE.")
    }
    return(out)
  }
  out <- split(seq_along(y), out)
  if (shuffle) {
    out <- lapply(out, .shuffle)
  }
  out
}

# Little helpers

# Save shuffling (even if x has length 1)
.shuffle <- function(x, ...) {
  x[sample.int(length(x), ...)]
}

# Efficient binning
.bin <- function(y, n_bins) {
  qu <- stats::quantile(y, seq(0, 1, length.out = n_bins + 1L), na.rm = TRUE)
  findInterval(y, unique(qu), rightmost.closed = TRUE)
}

# This this secret heart of splitTools
.smp_fun <- function(n, p) {
  sample(rep.int(seq_along(p), times = ceiling(p * n)), n)
}

# Fills empty partitions
.fill_empty_partitions <- function(fi, p) {
  counts <- tabulate(fi, nbins = length(p))
  empty <- which(counts == 0L)
  n_empty <- length(empty)

  if (n_empty == 0L) {
    return(fi)
  }

  message("Empty partition detected. Redistributing...")

  # Find positions of potential donors
  drop_random <- function(z) {
    m <- length(z) - 1L
    if (m >= 1L) sample(z, m)
  }
  positions <- split(seq_along(fi), fi)
  donors <- unlist(lapply(positions, drop_random), use.names = FALSE)
  n_donors <- length(donors)

  # Randomly select donors
  if (n_empty > n_donors) {
    message("Cannot fill all empty partitions.")
    n_empty <- n_donors
  }
  selected <- donors[sample.int(n_donors, n_empty)]

  # Replace donors by empty partition numbers
  fi[selected] <- empty[seq_len(n_empty)]
  fi
}
